hbase-server/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryEndpoint.java

```diff
@@ -0,0 +1,56 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.client;
+
+import java.util.List;
+import java.util.Optional;
+import org.apache.hadoop.hbase.HRegionLocation;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Define the necessary method for carrying {@code ClientMetaService}.
+ */
+@InterfaceAudience.Private
+public interface ConnectionRegistryEndpoint {
```

Comment:
Can't you just do, this doesn't work for any reason?

```
public interface ServerConnectionRegistry extends ConnectionRegistry {}

HRegionServer implements ServerConnectionRegistry {

  setupConnection(this);
}
```

Reply:
I think having two interfaces that look very much alike is only adding to the confusion of the code readers.

Reply:
This is a prior work for undoing HMaster extends HRegionServer. If HMaster does not extends HRegionServer, then it can not use RegionServerRegistry any more. So I agree with you that in the current code it is not necessary to do this refactoring, but I think it is also no big harm? If we could land this first, the final patch for HBASE-25288 could be smaller. But anyway, I respect your decision, if you do not think this is good to land, I’m OK that we just close this one and include this patch in the final patch for HBASE-25288.

Reply:
I agree that we should land this (thanks for your other patch btw). All I'm saying is perhaps we could do this refactor _without creating another new interface ConnectionRegistryEndpoint that looks exactly like ConnectionRegistry?_ (like I mentioned in my first comment, pasting again below). This seems simpler than your patch, right, you can use 'this' instance as the registry object in both HMaster and HRegionServer, don't have to construct something explicit? or did I miss something?

```
public interface ServerConnectionRegistry extends ConnectionRegistry {}

HMaster implements ServerConnectionRegistry {

 <Implement master based registry server side>
}

HRegionServer implements ServerConnectionRegistry {

<Implement RegionServerRegistry server side here>
}

```

Reply:
OK, got it. Let me have a try. Was wondering that returning a CompletableFuture at server side while there is no actual asynchronous in code is a bit strange, but seems no big real harm.

Reply:
Oh, I know why I introduce a special interface then, we need to use these methods in RSRpcServices and MasterRpcServices too, to return these information to connection registry running at client side.

It will be a bit strange that, we get a CompletableFuture which is not async actually...

WDYT? You can take a look at the code in RSRpcService.

https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java#L4092